MapboxCoreNavigation/RouteController.swift

```diff
@@ -605,6 +632,103 @@ extension RouteController: CLLocationManagerDelegate {
         checkForFasterRoute(from: location)
     }
     
+    func checkForTunnelIntersection(at location: CLLocation, for manager: CLLocationManager, distanceTraveled: CLLocationDistance) {
+        let currentIntersection = routeProgress.currentLegProgress.currentStepProgress.currentIntersection
+        let shouldAnimate = shouldEnableTunnelAnimation(at: location, for: manager, intersection: currentIntersection)
+        
+        if shouldAnimate {
+            enableTunnelAnimation(for: manager,
+                        routeProgress: routeProgress,
+                     distanceTraveled: distanceTraveled,
+                             callback: nil)
+        } else {
+            suspendTunnelAnimation(for: manager,
+                                    at: location,
+                              callback: nil)
+        }
+    }
+    
+    func shouldEnableTunnelAnimation(at location: CLLocation, for manager: CLLocationManager, intersection: Intersection?) -> Bool {
+        guard let currentIntersection = intersection else { return false }
+        
+        if let classes = currentIntersection.outletRoadClasses {
+            // Main conditions to enable simulated tunnel animation:
+            // - User location is within minimum tunnel entrance radius
+            // - OR Current intersection's road classes contain a tunnel AND when we receive series of bad GPS location updates
+            let upcomingIntersection = routeProgress.currentLegProgress.currentStepProgress.upcomingIntersection
+            if userWithinTunnelEntranceRadius(at: location, intersection: upcomingIntersection) ||
+               (classes.contains(.tunnel) && (manager is NavigationLocationManager && !location.isQualified))
+            {
+                return true
+            }
+        }
+
+        return false
+    }
+
+    func enableTunnelAnimation(for manager: CLLocationManager,
+                             routeProgress: RouteProgress,
+                          distanceTraveled: CLLocationDistance,
+                                  callback: RouteControllerSimulationCompletionBlock?) {
+       
+        guard !(manager is SimulatedLocationManager), animatedLocationManager == nil else { return }
```

Comment:
@vincethecoder Ideally we wouldn't have to check the type of class here. This check currently prevents us from replaying routes. 

Are there any other alternate ways we could bail early here?

Reply:
@bsudekum hmmmm... why would you want to replay routes here? Further  explanation would certainly help.